### PR TITLE
rgw/s3select: s3select output serialization

### PR DESCRIPTION
**Pull request by albin-antony**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_rest_s3.cc [hunk 1]
```diff
@@ -6613,6 +6636,11 @@ int RGWSelectObj_ObjStore_S3::handle_aws_cli_parameters(std::string& sql_query)
   extract_by_tag("QuoteEscapeCharacter", m_escape_char);
   extract_by_tag("CompressionType", m_compression_type);
   extract_by_tag("Enabled", m_enable_progress);
+  extract_by_tag("FieldDelimiter", output_column_delimiter);
```
Comment (galsalomon66) [inline src/rgw/rgw_rest_s3.cc:6639]: the tag "FieldDelimiter" is been used twice for  m_column_delimiter and output_column_delimiter

File: src/rgw/rgw_rest_s3.cc [hunk 2]
```diff
@@ -6613,6 +6636,11 @@ int RGWSelectObj_ObjStore_S3::handle_aws_cli_parameters(std::string& sql_query)
...
+  extract_by_tag("QuoteCharacter", output_quot);
+  extract_by_tag("QuoteEscapeCharacter", output_escape_char);
+  extract_by_tag("QuoteFields", output_quote_fields);
+  extract_by_tag("RecordDelimiter", output_row_delimiter);
```
Comment (galsalomon66) [inline src/rgw/rgw_rest_s3.cc:6643]: the tag "RecordDelimiter" is used twice for RecordDelimiter and m_row_delimiter

**Review by galsalomon66 — COMMENTED** [review]
please review the comments.
the tag extraction is not correct, upon extracting the new tags, the CSV object should be set accordingly.

File: qa/suites/rgw/verify/tasks/s3tests.yaml [hunk 3]
```diff
@@ -1,5 +1,5 @@
 tasks:
 - s3tests:
     client.0:
-      force-branch: ceph-master
+      force-branch: progress-stats
```
Comment (galsalomon66) [inline qa/suites/rgw/verify/tasks/s3tests.yaml:4]: no need for that 
it should be ceph-master

**Comment by galsalomon66** [comment]
please create an interface for testing the output serialization in the engine itself (s3select_example).

and also create an interface in s3test(boto3) for running queries with different output-serialization parameters.

File: src/rgw/rgw_rest_s3.cc [hunk 4]
```diff
@@ -6431,28 +6668,43 @@ int RGWSelectObj_ObjStore_S3::extract_by_tag(std::string tag_name, std::string&
...
 }
 
-int RGWSelectObj_ObjStore_S3::send_response_data(bufferlist& bl, off_t ofs, off_t len)
+int RGWSelectObj_ObjStore_S3::extract_output_by_tag(std::string tag_name, std::string& result)
 {
-  if (len == 0) {
-    return 0;
+  result = "";
+  std::string output_tag{"OutputSerialization"};
+  size_t _qo = m_s3select_query.find("<" + output_tag + ">", 0) + output_tag.size() + 2;
```
Comment (galsalomon66) [inline src/rgw/rgw_rest_s3.cc:6675]: its not correct to do addition operation with `std::string.find`, find may return `::npos`.
there are more code lines with such an operation that should be fixed.

the input-serialization and output-serialization are both using the same tag-names, it causes wrong assignments,
the current functionality is not fit for that.

File: src/rgw/rgw_rest_s3.h [hunk 5]
```diff
@@ -914,42 +914,135 @@ class s3select;
...
   std::unique_ptr<s3selectEngine::csv_object> m_s3_csv_object;
   std::string m_column_delimiter;
   std::string m_quot;
   std::string m_row_delimiter;
   std::string m_compression_type;
   std::string m_escape_char;
-  std::unique_ptr<char[]>  m_buff_header;
   std::string m_header_info;
   std::string m_sql_query;
+  std::string m_enable_progress;
```
Comment (galsalomon66) [inline src/rgw/rgw_rest_s3.h:1033]: the new data members should be initialized in the constructor. and also the "old" ones.
please keep the current convention, add m_ to data-member.

**Comment by galsalomon66** [comment]
please review the comments.
the extract_by_tag functions should be re-factor or replaced.

**Comment by galsalomon66** [comment]
please try the following(use CSV with comma delimiter).

aws s3api select-object-content --endpoint-url http://localhost:8000 --bucket <bucket-name> --key <key-name> --expression-type 'SQL' --input-serialization '{"CSV": {"RecordDelimiter" :"\n"}, "CompressionType": "NONE"}' --output-serialization '{"CSV": {"RecordDelimiter" :"\t","FieldDelimiter" :"="}}' --expression "select _1,_3 from s3object where _1 = _3;" "output.csv" && cat output.csv

File: src/rgw/rgw_rest_s3.cc (hunk 4 above)
Comment (albin-antony) [reply src/rgw/rgw_rest_s3.cc]: Thanks @galsalomon66. I have updated. It might need some refactoring. Tests are passing.

**Comment by galsalomon66** [comment]
running the following cause RGW to crash (it consumes a lot of memory before it crashes).

aws s3api select-object-content --endpoint-url http://localhost:8000 --bucket galsalomon-bucket --key 10krows_10col.csv --expression-type 'SQL' --input-serialization '{"CSV": {}, "CompressionType": "NONE"}' --output-serialization '{"CSV": {"QuoteFields": "ALWAYS","RecordDelimiter": "\n","FieldDelimiter": "="}}' --expression "select * from s3object where _1 = _2;" "output.csv" && cat output.csv

File: src/rgw/rgw_rest_s3.cc [hunk 6]
```diff
@@ -6338,47 +6509,109 @@ int RGWSelectObj_ObjStore_S3::run_s3select(const char* query, const char* input,
...
       csv.use_header_info=true;
     }
-
     m_s3_csv_object = std::unique_ptr<s3selectEngine::csv_object>(new s3selectEngine::csv_object(s3select_syntax.get(), csv));
   }
 
-  header_size = create_header_records(m_buff_header.get());
-  m_result.append(m_buff_header.get(), header_size);
-  m_result.append(PAYLOAD_LINE);
+  m_aws_response_handler.get()->init_response();
```
Comment (cbodley) [inline src/rgw/rgw_rest_s3.cc:6550]: you can use -> directly on unique_ptrs: `m_aws_response_handler->init_response()`

same below

**Comment by galsalomon66** [comment]
@albin-antony please close this PR

**Pull request closed without merging**